File: 05-Factory-Method/src/go/factory.go
package main

import "fmt"

// Creator interface — Go uses interface, not abstract class
// NOTE: Go can't have default methods on interfaces.
// We use a free function PlanDelivery() that takes the factory.

type LogisticsFactory interface {
	CreateTransport() Transport
}

// Free function with business logic (since Go interfaces can't have default methods)
func PlanDelivery(factory LogisticsFactory) {
	transport := factory.CreateTransport()
	fmt.Println("Planning delivery with:", transport.Name())
	transport.Deliver()
}

// Concrete Creators
type RoadLogistics struct{}
func (RoadLogistics) CreateTransport() Transport { return Truck{} }

type SeaLogistics struct{}
func (SeaLogistics) CreateTransport() Transport { return Ship{} }

type AirLogistics struct{}
func (AirLogistics) CreateTransport() Transport { return Airplane{} }

File: 05-Factory-Method/src/rust/factory.rs
use crate::transport::*;

// Creator trait — factory method + business logic
// NOTE: Rust uses trait with default method instead of abstract class
pub trait LogisticsFactory {
    fn create_transport(&self) -> Box<dyn Transport>;

    fn plan_delivery(&self) {
        let transport = self.create_transport();
        println!("Planning delivery with: {}", transport.name());
        transport.deliver();
    }
}

// Concrete Creators
pub struct RoadLogistics;
impl LogisticsFactory for RoadLogistics {
    fn create_transport(&self) -> Box<dyn Transport> { Box::new(Truck) }
}

pub struct SeaLogistics;
impl LogisticsFactory for SeaLogistics {
    fn create_transport(&self) -> Box<dyn Transport> { Box::new(Ship) }
}

pub struct AirLogistics;
impl LogisticsFactory for AirLogistics {
    fn create_transport(&self) -> Box<dyn Transport> { Box::new(Airplane) }
}
